File: alxSearch/src/search1.cpp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <regex>
#include <unordered_map>

namespace fs = std::filesystem;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <directory>\n";
        return 1;
    }

    fs::path dir = argv[1];
    std::regex token_regex(R"([A-Z][a-z]*)"); // Worttoken-Regel

    std::unordered_map<std::string, int> global_count;
    std::vector<std::tuple<std::string, std::string, int, int>> csv_rows;

    for (auto& p: fs::recursive_directory_iterator(dir)) {
        if (!fs::is_regular_file(p)) continue;
        auto size = fs::file_size(p);
        if (size > 500 * 1024) continue; // >500 KB

        std::ifstream file(p);
        if (!file) continue;

        std::unordered_map<std::string, int> file_count;
        std::string line;
        while (std::getline(file, line)) {
            for (std::sregex_iterator it(line.begin(), line.end(), token_regex), end_it; it != end_it; ++it) {
                std::string token = it->str();
                ++file_count[token];
                ++global_count[token];
            }
        }

        for (auto& [token, count] : file_count) {
            csv_rows.emplace_back(token, p.filename().string(), 0, count); // 0 als Platzhalter für global_count
        }
    }

    // Update globale Häufigkeit in CSV
    for (auto& row : csv_rows) {
        std::get<2>(row) = global_count[std::get<0>(row)];
    }

    // CSV schreiben
    std::ofstream csv("worttokens.csv");
    csv << "Token,Dateiname,HäufigkeitGesamt,HäufigkeitDatei\n";
    for (auto& row : csv_rows) {
        csv << std::get<0>(row) << "," 
            << std::get<1>(row) << ","
            << std::get<2>(row) << ","
            << std::get<3>(row) << "\n";
    }

    std::cout << "CSV erzeugt: worttokens.csv\n";
}

File: alxSearch/src/search1.rs
use std::collections::HashMap;
use std::fs;
use std::io::{self, BufRead};
use regex::Regex;
use std::path::Path;

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <directory>", args[0]);
        return Ok(());
    }

    let dir = &args[1];
    let token_regex = Regex::new(r"[A-Z][a-z]*").unwrap();

    let mut global_count: HashMap<String, usize> = HashMap::new();
    let mut csv_rows: Vec<(String, String, usize, usize)> = Vec::new();

    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() {
            if fs::metadata(&path)?.len() > 500 * 1024 { continue; }

            let file = fs::File::open(&path)?;
            let reader = io::BufReader::new(file);

            let mut file_count: HashMap<String, usize> = HashMap::new();
            for line in reader.lines() {
                let line = line?;
                for cap in token_regex.captures_iter(&line) {
                    let token = cap[0].to_string();
                    *file_count.entry(token.clone()).or_insert(0) += 1;
                    *global_count.entry(token.clone()).or_insert(0) += 1;
                }
            }

            for (token, count) in file_count {
                csv_rows.push((token, path.file_name().unwrap().to_string_lossy().to_string(), 0, count));
            }
        }
    }

    // Update global count
    for row in csv_rows.iter_mut() {
        row.2 = global_count[&row.0];
    }

    let mut wtr = csv::Writer::from_path("worttokens.csv")?;
    wtr.write_record(&["Token", "Dateiname", "HäufigkeitGesamt", "HäufigkeitDatei"])?;
    for row in csv_rows {
        wtr.write_record(&[row.0, row.1, row.2.to_string(), row.3.to_string()])?;
    }
    wtr.flush()?;

    println!("CSV erzeugt: worttokens.csv");
    Ok(())
}
